{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import glob\n",
    "from collections import defaultdict\n",
    "from itertools import combinations, product\n",
    "\n",
    "project_dir = \"\"\n",
    "contig_extension = \"\"\n",
    "strain_list_file = \"\"\n",
    "output_prefix = \"\"\n",
    "pop_infile_name = \"\"\n",
    "\n",
    "with open(\"phybreak_parameters.txt\",\"r\") as parameter_file:\n",
    "    for line in parameter_file:\n",
    "        line = line.strip().split(\" = \")\n",
    "        if len(line) > 1:\n",
    "            if line[0] == \"project_dir\":\n",
    "                project_dir = line[1].split(\" #\")[0]\n",
    "            elif line[0] == \"contig_extension\":\n",
    "                contig_extension = line[1].split(\" #\")[0]\n",
    "            elif line[0] == \"strain_list_file\":\n",
    "                strain_list_file = line[1].split(\" #\")[0]\n",
    "            elif line[0] == \"pop_infile_name\":\n",
    "                pop_infile_name = line[1].split(\" #\")[0]\n",
    "            elif line[0] == \"output_prefix\":\n",
    "                output_prefix = line[1].split(\" #\")[0]\n",
    "\n",
    "phy_split_dir = input_dir + \"phy_split/\"\n",
    "phy_prefix = output_prefix + \".core\"\n",
    "\n",
    "# Define some functions\n",
    "def count_divs(s1, s2):\n",
    "    '''\n",
    "    Given two strings counts the number of differences between them\n",
    "    '''\n",
    "    d = 0\n",
    "    for b1, b2 in zip(s1, s2):\n",
    "        if b1 != b2:\n",
    "            d += 1\n",
    "    return d * 1.0 / len(s1)\n",
    "\n",
    "def calc_pop_div(pop, s_dict):\n",
    "    '''\n",
    "    Given a list of strains and a dictionary of sequences, calculates pi,\n",
    "    the average nucleotide diversity between pairs of strains\n",
    "    '''\n",
    "    \n",
    "    # pop is just a list of strains in the population of interest\n",
    "    # s dict is a dictionary that maps strains/genomes to sequences\n",
    "    aves = []\n",
    "    for strain1, strain2 in combinations(pop, 2):\n",
    "        if strain1 in s_dict.keys() and strain2 in s_dict.keys():\n",
    "            aves.append(count_divs(s_dict[strain1], s_dict[strain2]))\n",
    "    return(np.average(aves))\n",
    "\n",
    "def calc_inter_div(pop_1, pop_2, s_dict):\n",
    "    '''\n",
    "    Given two lists of strains representing two populations\n",
    "    and a dictionary of sequences, calculate pi between the \n",
    "    two populations \n",
    "    '''\n",
    "    aves = []\n",
    "    for strain1, strain2 in product(pop_1, pop_2):\n",
    "        if strain1 in s_dict.keys() and strain2 in s_dict.keys():\n",
    "            aves.append(count_divs(s_dict[strain1], s_dict[strain2]))\n",
    "    return(np.average(aves))\n",
    "\n",
    "\n",
    "# inputs\n",
    "\n",
    "# File that maps trees to alignment positions\n",
    "info_file = input_dir + output_prefix + '.core.phyml_tree_info.txt'\n",
    "\n",
    "# ClusterPop assignments\n",
    "pop_file = '../../Data/Clusters/rumino_0.000355362.txt.cluster.tab.txt'\n",
    "\n",
    "# Alignment of each tree\n",
    "individual_alignments = phy_split_dir + \"*window.phy\"\n",
    "tree_aligns = glob.glob(individual_alignments)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "info = pd.read_table(info_file, index_col='tree')\n",
    "pops = pd.read_table(pop_file, index_col='Strain')\n",
    "pop_assignments = {}\n",
    "\n",
    "# Have to rename the strains so that they match the sequence files\n",
    "for pop, pop_df in pops.groupby('Cluster_ID'):\n",
    "    pop_assignments[str(pop)] = [strain.split('--')[0].replace('.','').replace('_', '') for strain in list(pop_df.index)]\n",
    "    \n",
    "all_strains = []\n",
    "for strain_list in pop_assignments.values():\n",
    "    all_strains += strain_list\n",
    "\n",
    "columns = []\n",
    "for pop, strains in pop_assignments.items():\n",
    "    popID = 'Pop' + pop.split('.')[-1]\n",
    "    columns.append(popID)\n",
    "    columns.append(popID + ' v All')\n",
    "    \n",
    "for pop1, pop2 in combinations(pop_assignments, 2):\n",
    "    popID1 = 'Pop' + pop1.split('.')[-1]\n",
    "    popID2 = 'Pop' + pop2.split('.')[-1]\n",
    "    columns.append(popID1 + ' v ' + popID2)\n",
    "    \n",
    "pi_df = pd.DataFrame(index=range(1, len(blocks) + 1), columns=columns)\n",
    "\n",
    "for tree_align in tree_aligns:\n",
    "    # Get number of tree\n",
    "    treenum = int(tree_align.split('_')[-1].split('.phy')[0])\n",
    "    \n",
    "    all_seqs = {line.strip().split()[0].split('_')[0]: line.split()[1] for line in open(block) if line[0] != ' '}\n",
    "    for pop, strains in pop_assignments.items():\n",
    "        popID = 'Pop' + pop.split('.')[-1]\n",
    "        \n",
    "        # Diversity within a population\n",
    "        intra_pop_div = calc_pop_div(strains, all_seqs)\n",
    "        \n",
    "        # Diversity between focus population and all other strains\n",
    "        pop_v_all = calc_inter_div(strains, set(all_strains) - set(strains), all_seqs)\n",
    "        \n",
    "        pi_df.loc[blocknum, popID] = intra_pop_div\n",
    "        pi_df.loc[blocknum, popID + ' v All'] = pop_v_all\n",
    "        \n",
    "    for pop1, pop2 in combinations(pop_assignments, 2):\n",
    "        # Diversity between each pair of populations\n",
    "        popID1 = 'Pop' + pop1.split('.')[-1]\n",
    "        popID2 = 'Pop' + pop2.split('.')[-1]\n",
    "        strains1 = pop_assignments[pop1]\n",
    "        strains2 = pop_assignments[pop2]\n",
    "        inter_div = calc_inter_div(strains1, strains2, all_seqs)\n",
    "        pi_df.loc[blocknum, popID1 + ' v ' + popID2] = pop_v_all\n",
    "\n",
    "pi_df.to_csv('pi_per_block.csv')\n",
    "outdf = pi_df.join(info)\n",
    "outdf.index.name='tree'\n",
    "outdf.to_csv('pi_join_info.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
