{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "import glob\n",
    "import pandas as pd\n",
    "from itertools import combinations, product\n",
    "import numpy as np\n",
    "from scipy import stats \n",
    "\n",
    "# Input files\n",
    "pop_map = '' # ClusterPop Clusters\n",
    "msa_file = '' # Concatenated whole genome alignment from phybreak script 2\n",
    "pop0_phybreak_output = '' # file that indicates for each tree, whether population I is monophyletic and what percentage of branch length lies within population I\n",
    "pop1_phybreak_output = '' # same as above but for population II\n",
    "core_sweep_pop_I_outfile = '' #\n",
    "core_sweep_pop_II_outfile = '' #\n",
    "\n",
    "p1 = 0.0138272139869 # Average nucleotide diversity of population I\n",
    "p0 = 0.00521122600799 # Average nucleotide diversity of population II\n",
    "alpha = 0.95 # Cutoff for determining whether a putative sweep region is significantly lower in diversity than expected\n",
    "\n",
    "total_alignment = {s.id : str(s.seq) for s in SeqIO.parse(msa_file, 'fasta')}\n",
    "\n",
    "df = pd.read_table(pop_map)\n",
    "pop0 = [strain.replace('_', '').replace('.','').replace('--', '_') for strain in df[df['Cluster_ID'] == 0].Strain]\n",
    "pop1 = [strain.replace('_', '').replace('.','').replace('--', '_') for strain in df[df['Cluster_ID'] == 0.1].Strain]\n",
    "pop2 = [strain.replace('_', '').replace('.','').replace('--', '_') for strain in df[df['Cluster_ID'] == 0.2].Strain]\n",
    "\n",
    "# div_df is the dataframe of nucleotide diversity per block\n",
    "div_df = pd.read_csv('pi_join_info.csv', index_col=0)\n",
    "div_df['block'] = div_df.index\n",
    "div_df['Length'] = div_df.end - div_df.start\n",
    "div_df['Midpoint'] = div_df.start + (div_df.end - div_df.start) // 2\n",
    "div_df['Start'] = div_df['start']\n",
    "div_df['End'] = div_df['end']\n",
    "div_df = div_df.sort_values('Midpoint')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def concatenate_windows(sweepdf, length_cutoff=1000):\n",
    "    i = 0\n",
    "    final_df = pd.DataFrame()\n",
    "    current_end = None\n",
    "    blocks = []\n",
    "    for index, row in sweepdf.iterrows():\n",
    "        if not current_end:\n",
    "            current_end = row['End']\n",
    "            blocks.append(index)\n",
    "        elif row['Start'] <= current_end:\n",
    "            blocks.append(index)\n",
    "\n",
    "        else: \n",
    "            if len(blocks) > 0:\n",
    "                if sweepdf.loc[max(blocks), 'End'] - sweepdf.loc[min(blocks), 'Start'] >= length_cutoff:\n",
    "                    block_begin = min(blocks)\n",
    "                    block_end = max(blocks)\n",
    "                    final_df.loc[i, 'Start'] = sweepdf.loc[min(blocks), 'Start']\n",
    "                    final_df.loc[i, 'End'] = sweepdf.loc[max(blocks), 'End']\n",
    "                    final_df.loc[i, 'Start tree'] = min(blocks)\n",
    "                    final_df.loc[i, 'End tree'] = max(blocks)\n",
    "                    i += 1\n",
    "            blocks = []\n",
    "        current_end = row['End']\n",
    "        \n",
    "    if len(blocks) > 0:\n",
    "        if sweepdf.loc[max(blocks), 'End'] - sweepdf.loc[min(blocks), 'Start'] >= length_cutoff:\n",
    "            block_begin = min(blocks)\n",
    "            block_end = max(blocks)\n",
    "            final_df.loc[i, 'Start'] = sweepdf.loc[min(blocks), 'Start']\n",
    "            final_df.loc[i, 'End'] = sweepdf.loc[max(blocks), 'End']\n",
    "            final_df.loc[i, 'Start tree'] = min(blocks)\n",
    "            final_df.loc[i, 'End tree'] = max(blocks)\n",
    "\n",
    "    final_df['Midpoint'] = final_df.Start + ((final_df.End - final_df.Start) // 2)\n",
    "    return final_df\n",
    "\n",
    "def calc_all_divs(concatenated_positions,\n",
    "                  population,\n",
    "                  pi,\n",
    "                  alpha):\n",
    "    finaldf = pd.DataFrame(index=concatenated_positions.index, columns=['Start',\n",
    "                                                                        'End',\n",
    "                                                                        'Start tree',\n",
    "                                                                        'End tree',\n",
    "                                                                        'pop_pi'])\n",
    "    for i in concatenated_positions.index:\n",
    "        start = int(concatenated_positions.loc[i, 'Start'])\n",
    "        end = int(concatenated_positions.loc[i, 'End'])\n",
    "        seqs = {}\n",
    "        for strain, s in total_alignment.items():\n",
    "            seqs[strain] = s[start: end]\n",
    "        \n",
    "        pop_div = calc_pop_div(population, seqs)\n",
    "\n",
    "        finaldf.loc[i, ] = [start,\n",
    "                            end,\n",
    "                            concatenated_positions.loc[i, 'Start tree'],\n",
    "                            concatenated_positions.loc[i, 'End tree'],\n",
    "                            pop_div]\n",
    "        \n",
    "    finaldf['Length'] = finaldf.End - finaldf.Start\n",
    "    finaldf['Midpoint'] = finaldf.Start + ((finaldf.End - finaldf.Start) // 2)\n",
    "    calculate_ci(finaldf, 'pop_pi', alpha, pi)\n",
    "    return finaldf\n",
    "    \n",
    "def count_divs(s1, s2):\n",
    "    d = 0\n",
    "    for b1, b2 in zip(s1, s2):\n",
    "        if b1 != b2:\n",
    "            d += 1\n",
    "    return d * 1.0 / len(s1)\n",
    "\n",
    "def calc_pop_div(pop, s_dict):\n",
    "    aves = []\n",
    "    for strain1, strain2 in combinations(pop, 2):\n",
    "        if strain1 in s_dict.keys() and strain2 in s_dict.keys():\n",
    "            aves.append(count_divs(s_dict[strain1], s_dict[strain2]))\n",
    "    return(np.average(aves))\n",
    "\n",
    "def calculate_ci(df, pop_name, alpha, p_pop):\n",
    "    low, high = stats.binom.interval(alpha, list(df.Length), p_pop)\n",
    "    df['ci_low'] = low / df.Length\n",
    "    df['ci_high'] = high / df.Length\n",
    "\n",
    "def passes_phylo_criteria(full_df, cutoff, focus_pop):\n",
    "    return (full_df.focus < cutoff) & (full_df.monophy == 1) & ((full_df[focus_pop] < full_df.ci_low) | (full_df[focus_pop]==0))\n",
    "\n",
    "def find_sweeps(phybreak_df,\n",
    "                pi_info_df,\n",
    "                pop_pi, alpha,\n",
    "                pop_name,\n",
    "                pop_list):\n",
    "    full_df = pd.merge(pi_info_df, phybreak_df)\n",
    "    full_df.index = full_df.block\n",
    "    calculate_ci(full_df, pop_name, alpha, pop_pi)\n",
    "    \n",
    "    # Finds cutoff based on fraction of branch length within the focus population\n",
    "    cutoff = np.percentile(full_df.focus, 5)\n",
    "    trees_passing_phy_criteria = full_df[passes_phylo_criteria(full_df, cutoff, pop_name)]\n",
    "    \n",
    "    concat = concatenate_windows(trees_passing_phy_criteria, length_cutoff=500)\n",
    "    concat = calc_all_divs(concat, pop_list, pop_pi, alpha)\n",
    "    for col in concat:\n",
    "        concat[col] = concat[col].astype(float)\n",
    "    return (full_df, concat)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "pop_0_phybreak = pd.read_table(pop0_phybreak_output)\n",
    "plotdf_pop0, concat_pop0 = find_sweeps(pop_0_phybreak, div_df, p0, alpha, 'Pop0', pop0)\n",
    "concat_pop0.to_csv(core_sweep_pop_I_outfile, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_1_phybreak = pd.read_table(pop1_phybreak_output)\n",
    "plotdf_pop1, concat_pop1 = find_sweeps(pop_1_phybreak, div_df, p1, alpha, 'Pop1', pop1)\n",
    "concat_pop1.to_csv(core_sweep_pop_II_outfile, index=None)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
